File: src/FlatF.hs
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE NoMonomorphismRestriction #-}

-- * Demonstrating `non-compositional', context-sensitive processing
-- * The final style
-- * Flatten the additions

module FlatF where

import Intro2 hiding (main)
import PushNegF as Neg hiding (main,Ctx)

-- We are going to write flata as an interpreter.
-- After all, that's all we can do with an expression in a final form.

-- * The nested pattern-matching establishes a context:
-- * flata :: Exp -> Exp
-- * flata e@Lit{} = e
-- * flata e@Neg{} = e
-- * flata (Add (Add e1 e2) e3) = flata (Add e1 (Add e2 e3))
-- * flata (Add e1 e2)          = Add e1 (flata e2)

-- The processing is slightly different from push_neg, because we repeatedly
-- process the transformed expression (the last-but one clause)
-- The nested pattern-match again betrays the context-sensitivity.
-- The context in question is expression being the left child of addition.

-- * //
-- So, we define the context
-- (LCA e) means that the context (Add [] e), being the left immediate
-- child of addition whose right child is e. NonLCA is any other context.
data Ctx e = LCA e | NonLCA
-- On other words, LCA e3 represents the context of adding e3 _to the right_ of
-- the expression in focus.

instance ExpSYM repr => ExpSYM (Ctx repr -> repr) where
    lit n NonLCA   = lit n
    lit n (LCA e)  = add (lit n) e
    neg e NonLCA   = neg (e NonLCA)
    neg e (LCA e3) = add (neg (e NonLCA)) e3 -- assume only lits are negated
    add e1 e2 ctx  = e1 (LCA (e2 ctx))

-- The last clause expresses the reassociation-to-the-right
-- *	C[Add e1 e2] -> Add e1 C[e2]
-- Recall that after the negations are pushed down, expressions are
-- described by the following grammar
-- *    e ::= factor | add e e
-- *    factor ::= int | neg int
-- That is, C ::= [] | Add [] e

-- Keep in mind the processing is done bottom-up!

-- A general approach is to use continuations (or monads), 
-- or to encode the Ctx with zippers 
-- (the defunctionalized continuations, in the present
-- case).
-- The Ctx data type above is tuned to the present example,
-- keeping only the data we need (e.g., we aren't interested
-- in the context of Neg or of the addition on the left, 
-- and so we don't represent these cases in Ctx).


-- Exercise: there are several clauses where lit, neg, or add
-- appears on both sides of the definition. 
-- In which clause add is being used recursively?


-- The `interpreter' for flattening
flata :: (Ctx repr -> repr) -> repr
flata e = e NonLCA

--norm :: (Neg.Ctx -> Ctx c -> c) -> c
norm = flata . Neg.push_neg

-- Use our sample term
-- We make it a bit complex
tf3 = (add tf1 (neg (neg tf1)))

tf3_view = view tf3
-- "((8 + (-(1 + 2))) + (-(-(8 + (-(1 + 2))))))"

tf3_eval = eval tf3
-- 10

-- The normalized expression can be evaluated with any interpreter

tf3_norm = norm tf3

tf3_norm_view = view tf3_norm
-- "(8 + ((-1) + ((-2) + (8 + ((-1) + (-2))))))"

-- The result of the standard evaluation (the `meaning') is preserved

tf3_norm_eval = eval tf3_norm
-- 10

tf4 = add t (neg t) where t = (add (add (lit 1) (lit 2)) (lit 3))

tf4_view = view tf4
-- "(((1 + 2) + 3) + (-((1 + 2) + 3)))"

tf4_eval = eval tf4
-- 0

tf4_norm = norm tf4

tf4_norm_view = view tf4_norm
-- "(1 + (2 + (3 + ((-1) + ((-2) + (-3))))))"

-- The result of the standard evaluation (the `meaning') is preserved

tf4_norm_eval = eval tf4_norm
-- 0

main = do
       print tf3_view
       print tf3_eval
       print tf3_norm_view
       print tf3_norm_eval
       if tf3_eval == tf3_norm_eval then return ()
	  else error "Normalization"
       -- normalizing a normal form does not change it
       if (view . norm $ tf3_norm) == tf3_norm_view then return ()
	  else error "Normalization"
       print tf4_view
       print tf4_eval
       print tf4_norm_view
       print tf4_norm_eval
       if tf4_eval == tf4_norm_eval then return ()
	  else error "Normalization"
       if (view . norm $ tf4_norm) == tf4_norm_view then return ()
	  else error "Normalization"

File: src/scala/ttfi/FlatF.scala
package ttfi

import Intro2._
import PushNegF._

trait FlatF {
  sealed trait Ctx[+E]
  case class LCA[E](e: E) extends Ctx[E]
  case object NonLCA extends Ctx[Nothing]

  implicit def ctxExpSYM[R: ExpSYM]: ExpSYM[Ctx[R] => R] = new ExpSYM[Ctx[R] => R] {
    def lit = n => {
      case NonLCA => ExpSYM[R].lit(n)
      case LCA(e) => ExpSYM[R].add(ExpSYM[R].lit(n), e)
    }
    def neg = e => {
      case NonLCA  => ExpSYM[R].neg(e(NonLCA))
      case LCA(e3) => ExpSYM[R].add(ExpSYM[R].neg(e(NonLCA)), e3)
    }
    def add = (e1, e2) => ctx => e1(LCA(e2(ctx)))
  }

  def flata[R]: (Ctx[R] => R) => R =
    e => e(NonLCA)

  def norm[R]: (PushNegF.Ctx => Ctx[R] => R) => R = flata[R] compose pushNeg[Ctx[R] => R]
  // norm1(e[R]) = norm(e[PushNegF.Ctx => Ctx[R] => R])
  def tf3[R](implicit R: ExpSYM[R]): R = {
    import R._
    add(tf1[R], neg(neg(tf1[R])))
  }

  val tf3_view = view(tf3[String])
  val tf3_eval = eval(tf3[Int])

  def tf3_norm[R: ExpSYM] = norm(tf3[PushNegF.Ctx => Ctx[R] => R])

  val tf3_norm_view = view(tf3_norm[String])
  val tf3_norm_eval = eval(tf3_norm[Int])

  def tf4[R](implicit R: ExpSYM[R]) = {
    import R._
    def t = add(add(lit(1), lit(2)), lit(3))
    add(t, neg(t))
  }

  val tf4_view = view(tf4[String])
  val tf4_eval = eval(tf4[Int])

  def tf4_norm[R: ExpSYM] = norm(tf4[PushNegF.Ctx => Ctx[R] => R])

  val tf4_norm_view = view(tf4_norm[String])
  val tf4_norm_eval = eval(tf4_norm[Int])

}
object FlatF extends FlatF
object FlatFMain extends Main with FlatF {

  println(tf3_view)
  println(tf3_eval)
  println(tf3_norm_view)
  println(tf3_norm_eval)
  require(tf3_eval == tf3_norm_eval, "Normalization")
  // normalizing a normal form does not change it
  require(view(norm(tf3_norm[PushNegF.Ctx => Ctx[String] => String])) == tf3_norm_view, "Normalization")
  println(tf4_view)
  println(tf4_eval)
  println(tf4_norm_view)
  println(tf4_norm_eval)
  require(tf4_eval == tf4_norm_eval, "Normalization")
  require(view(norm(tf4_norm[PushNegF.Ctx => Ctx[String] => String])) == tf4_norm_view, "Normalization")
}
